File: go/clients/openai_usage_conformance_test.go
package clients_test

import (
	"encoding/json"
	"os"
	"testing"

	"github.com/DexwoxBusiness/dexcost-sdk/go/clients"
)

type openAIUsageValidCase struct {
	ID       string                 `json:"id"`
	Usage    map[string]interface{} `json:"usage"`
	Expected map[string]int         `json:"expected"`
}

type openAIUsageInvalidCase struct {
	ID            string                 `json:"id"`
	Usage         map[string]interface{} `json:"usage"`
	ExpectedError string                 `json:"expected_error"`
}

func TestSharedOpenAIUsageConformance(t *testing.T) {
	raw, err := os.ReadFile("../../fixtures/openai_usage_conformance.json")
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var fixture struct {
		ValidCases   []openAIUsageValidCase   `json:"valid_cases"`
		InvalidCases []openAIUsageInvalidCase `json:"invalid_cases"`
	}
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}

	for _, testCase := range fixture.ValidCases {
		t.Run(testCase.ID, func(t *testing.T) {
			usage, err := clients.NormalizeOpenAIUsage(testCase.Usage)
			if err != nil {
				t.Fatalf("normalize: %v", err)
			}
			actual := map[string]int{
				"total_input_tokens":       usage.TotalInputTokens,
				"input_tokens":             usage.InputTokens,
				"cache_read_input_tokens":  usage.CacheReadInputTokens,
				"cache_write_input_tokens": usage.CacheWriteInputTokens,
				"total_output_tokens":      usage.TotalOutputTokens,
				"output_tokens":            usage.OutputTokens,
				"reasoning_output_tokens":  usage.ReasoningOutputTokens,
			}
			for key, expected := range testCase.Expected {
				if actual[key] != expected {
					t.Errorf("%s = %d, want %d", key, actual[key], expected)
				}
			}
		})
	}

	for _, testCase := range fixture.InvalidCases {
		t.Run(testCase.ID, func(t *testing.T) {
			_, err := clients.NormalizeOpenAIUsage(testCase.Usage)
			if err == nil || err.Error() != testCase.ExpectedError {
				t.Fatalf("error = %v, want %q", err, testCase.ExpectedError)
			}
		})
	}
}

File: typescript/tests/openai-usage-conformance.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";
import { normalizeOpenAIUsage } from "../src/instruments/openai-usage.js";

interface ValidCase {
  id: string;
  usage: Record<string, unknown>;
  expected: Record<string, number>;
}

interface InvalidCase {
  id: string;
  usage: Record<string, unknown>;
  expected_error: string;
}

const fixture = JSON.parse(readFileSync(
  new URL("../../fixtures/openai_usage_conformance.json", import.meta.url),
  "utf8",
)) as { valid_cases: ValidCase[]; invalid_cases: InvalidCase[] };

function snakeCaseUsage(usage: ReturnType<typeof normalizeOpenAIUsage>): Record<string, number> {
  return {
    total_input_tokens: usage.totalInputTokens,
    input_tokens: usage.inputTokens,
    cache_read_input_tokens: usage.cacheReadInputTokens,
    cache_write_input_tokens: usage.cacheWriteInputTokens,
    total_output_tokens: usage.totalOutputTokens,
    output_tokens: usage.outputTokens,
    reasoning_output_tokens: usage.reasoningOutputTokens,
  };
}

describe("shared OpenAI usage conformance", () => {
  for (const testCase of fixture.valid_cases) {
    it(testCase.id, () => {
      expect(snakeCaseUsage(normalizeOpenAIUsage(testCase.usage))).toEqual(testCase.expected);
    });
  }

  for (const testCase of fixture.invalid_cases) {
    it(testCase.id, () => {
      expect(() => normalizeOpenAIUsage(testCase.usage)).toThrow(testCase.expected_error);
    });
  }

  it("keeps DeepSeek top-level cache-hit tokens disjoint", () => {
    expect(normalizeOpenAIUsage({
      prompt_tokens: 20,
      completion_tokens: 10,
      prompt_cache_hit_tokens: 4,
      prompt_cache_miss_tokens: 16,
    })).toEqual({
      totalInputTokens: 20,
      inputTokens: 16,
      cacheReadInputTokens: 4,
      cacheWriteInputTokens: 0,
      totalOutputTokens: 10,
      outputTokens: 10,
      reasoningOutputTokens: 0,
    });
  });
});
